server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -871,30 +882,78 @@ public void onClusterStateUnchanged(ClusterState clusterState) {
-            try (ThreadContext.StoredContext ignored = task.threadContextSupplier.get()) {
+        void onPublishFailure(FailedToCommitClusterStateException e) {
+            if (publishListener == null) {
+                assert failure != null;
+                return;
+            }
+            try (ThreadContext.StoredContext ignored = updateTask.threadContextSupplier.get()) {
                 publishListener.onFailure(e);
             } catch (Exception inner) {
                 inner.addSuppressed(e);
                 logger.error("exception thrown by listener notifying of failure", inner);
             }
         }
 
-        public ContextPreservingAckListener getContextPreservingAckListener() {
-            return task.wrapInTaskContext(clusterStateAckListener);
+        ContextPreservingAckListener getContextPreservingAckListener() {
+            assert incomplete() == false;
+            return updateTask.wrapInTaskContext(clusterStateAckListener);
+        }
+
+        @Override
+        public String toString() {
+            return "TaskContextImpl[" + updateTask.getTask() + "]";
         }
     }
 
-    private List<NonFailedTask> getNonFailedTasks(TaskInputs taskInputs, ClusterTasksResult<ClusterStateTaskListener> clusterTasksResult) {
-        return taskInputs.updateTasks.stream().flatMap(updateTask -> {
-            assert clusterTasksResult.executionResults().containsKey(updateTask.getTask()) : "missing " + updateTask;
-            final ClusterStateTaskExecutor.TaskResult taskResult = clusterTasksResult.executionResults().get(updateTask.getTask());
-            if (taskResult.isSuccess()) {
-                return Stream.of(new NonFailedTask(updateTask, taskResult.publishListener(), taskResult.clusterStateAckListener()));
-            } else {
-                return Stream.of();
+    private record BatchExecutionResult(ClusterState resultingState, List<ExecutionResult<ClusterStateTaskListener>> executionResults) {}
+
+    private BatchExecutionResult executeTasks(TaskInputs taskInputs, ClusterState previousClusterState) {
+        final var batchExecutionResult = innerExecuteTasks(taskInputs, previousClusterState);
+        if (previousClusterState != batchExecutionResult.resultingState()
+            && previousClusterState.nodes().isLocalNodeElectedMaster()
+            && (batchExecutionResult.resultingState().nodes().isLocalNodeElectedMaster() == false)) {
+            throw new AssertionError("update task submitted to MasterService cannot remove master");
+        }
+        assert assertAllTasksComplete(batchExecutionResult.executionResults());
+        return batchExecutionResult;
+    }
+
+    private static boolean assertAllTasksComplete(List<ExecutionResult<ClusterStateTaskListener>> executionResults) {
+        for (final var executionResult : executionResults) {
+            assert executionResult.incomplete() == false : "missing result for " + executionResult;
+        }
```

Comment:
```suggestion
        assert executionResults.stream().allMathch(executionResult -> executionResult.incomplete() == false) : "Missing result in " + executionResults;
```

To avoid the entire loop

Reply:
Actually this is already called from assert, disregard above comment

Reply:
Yeah also I'd rather just report the first missing result in the message rather than all of them. FWIW `noneMatch` would be nicer here:

```suggestion
        assert executionResults.stream().noneMatch(ExecutionResult::incomplete);
```

Reply:
(also `stream().allMatch()` and friends don't actually avoid any loops - there's still a loop in there somewhere, and a bunch more objects to allocate too)

Reply:
Sure, I meant the loop outside of assertion before realizing that the entire method is called from another assertion